### PR TITLE
add graphql api

### PR DESCRIPTION
**Pull request by CircArgs**
This PR seeks to address https://github.com/DataJunction/datajunction/issues/154

Using [strawberry](https://strawberry.rocks/) wrapping the sqlmodel pydantic models used in the rest api it gives a graphql api accessible with the example at `localhost:8000/graphql`

Here are some examples

```graphql
{
  getDatabases {
    id
    queries{
      id
      submittedQuery
    }
    name
    description
    URI
    createdAt
    extraParams
    cost
    tables {
      id
      databaseId
      node {
        id
        type
        tables{
          id
          columns {
            id
          }
        }
      }
    }
  }
}
```

```graphql
mutation{
  submitQuery(
    createQuery: {databaseId: 2, submittedQuery: "select count(*) from comments"}
  ) {
    id
    submittedQuery
    executedQuery
    results{
      _Root__{
        rows
        rowCount
      }
    }
  }
}
```

```graphql
{
  readMetrics{
    name
    id
  }
}
```

```graphql
{
  readMetricsData(nodeId:4){
    id
  }
}
```

```graphql
{
  readMetricsSql(nodeId:4){
    sql
    databaseId
  }
}
```

```graphql
{
  getNodes {
    parents {
      type
      id
      name
    }
    children {
      type
      id
      name
    }
  }
}
```

### CODE REVIEW

**Comment by samredai** [comment]
_request_
```gql
{
  readMetricsSql(nodeId:10, d: "basic.transform.country_agg.country", f: "basic.transform.country_agg.country='Brazil'"){
    sql
    databaseId 
  }
}
```
_response_
```json
{
  "data": {
    "readMetricsSql": {
      "sql": "SELECT sum(\"basic.transform.country_agg\".num_users) AS sum_1, \"basic.transform.country_agg\".country \nFROM (SELECT \"basic.source.users\".country AS country, count(\"basic.source.users\".id) AS num_users \nFROM (SELECT basic.dim_users.id AS id, basic.dim_users.full_name AS full_name, basic.dim_users.age AS age, basic.dim_users.country AS country, basic.dim_users.gender AS gender, basic.dim_users.preferred_language AS preferred_language \nFROM basic.dim_users) AS \"basic.source.users\" GROUP BY \"basic.source.users\".country) AS \"basic.transform.country_agg\" \nWHERE \"basic.transform.country_agg\".country = 'Brazil' GROUP BY \"basic.transform.country_agg\".country",
      "databaseId": 3
    }
  }
}
```
---
_request_
```gql
{
  readMetricsData(nodeId:10, d: "basic.transform.country_agg.country", f: "basic.transform.country_agg.country='Brazil'"){
    databaseId
    results {
      _Root__ {
        rows
      }
    }
  }
}
```
_response_
```json
{
  "data": {
    "readMetricsData": {
      "databaseId": 3,
      "results": {
        "_Root__": [
          {
            "rows": [
              [
                "Decimal('1')",
                "'Brazil'"
              ]
            ]
          }
        ]
      }
    }
  }
}
```
VERY cool! 😃

**Review by agorajek — CHANGES_REQUESTED** [review]
@CircArgs thank you so much for this PR! This plays very well into our vision of making DJ api available to a wide variety of clients. 

so_good

I just have left few minor comments in-line but after that I think we can merge it. I think our next problem to solve will be writing some good tests to make sure that both APIs (internal + graphQL) are kept in sync as we develop this system, but we can cross that bridge when we get there.

File: src/datajunction/api/graphql/table.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+
+@strawberry.experimental.pydantic.type(
+    model=_Table, fields=["id", "node_id", "database_id"]
+)
+class Table:
+    """
+    A table with data.
+
+    Nodes can data in multiple tables, in different databases.
```
[the remark this answers is not in the document]
Comment (CircArgs) [reply src/datajunction/api/graphql/table.py]: Done!

File: src/datajunction/api/main.py [hunk 2]
```diff
@@ -8,12 +8,14 @@
 # pylint: disable=unused-import
 
 import logging
+from webbrowser import get
```
[the remark this answers is not in the document]
Comment (CircArgs) [reply src/datajunction/api/main.py]: I don't really know where this came from 😕 maybe my ide. removed it though

File: src/datajunction/api/graphql/query.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+@strawberry.experimental.pydantic.type(model=_BaseQuery, all_fields=True)
+class BaseQuery:
+    """
+    Base class for query models.
+    """
+
+
+@strawberry.experimental.pydantic.type(model=_Query, all_fields=True)
+class Query_:
```
[the remark this answers is not in the document]
Comment (CircArgs) [reply src/datajunction/api/graphql/query.py:33]: This is the one I don't think I can implement at least directly. I think the underscore is necessary to differentiate the type from the Grapqhl root `Query` type

File: src/datajunction/api/graphql/database.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        "URI",
+        "read_only",
+        "async_",
+        "cost",
+        "created_at",
+        "updated_at",
+    ],
+)
+class Database:
+    """a source database"""
```
[the remark this answers is not in the document]
Comment (CircArgs) [reply src/datajunction/api/graphql/database.py]: Done!

File: src/datajunction/api/graphql/column.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
+"""
+Models for columns.
+"""
+
+from typing import TYPE_CHECKING, Optional, TypedDict
+
+import strawberry
+from sqlalchemy.sql.schema import Column as SqlaColumn
```
[the remark this answers is not in the document]
Comment (CircArgs) [reply src/datajunction/api/graphql/column.py]: Nope my bad

File: src/datajunction/api/graphql/query.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (agorajek) [reply src/datajunction/api/graphql/query.py:33]: I see. I don't have an opinion on a better option then, but maybe @betodealmeida has.

**Review by agorajek — APPROVED** [review]

**Pull request merged**